### DISCUSSION TITLE
0.13.0-beta1 has been released!

### DISCUSSION THREAD

**Discussion by mihe** [godot-jolt/godot-jolt#888] [Announcements]
> [!WARNING]
> This release has been superseded by [0.13.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.13.0-stable). Please use that one instead, as it includes some important bug fixes.

This release only supports **Godot 4.3** (starting with [4.3-beta1](https://godotengine.org/article/dev-snapshot-godot-4-3-beta-1/)).

Note that this is a **beta** release, and as such is not available through the Godot Asset Library. It has also not been very thoroughly tested, so please report any issues you may run into, especially when dealing with some of the bigger changes, like the support for `SoftBody3D`.

Expect version 0.13.0-stable to be out some time around the release of Godot 4.3-stable.

## Changelog

These are the notable changes that have been made since [0.12.0-stable](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.12.0-stable) was released. You can also find a list of all the commits [here](https://github.com/godot-jolt/godot-jolt/compare/v0.12.0-stable...v0.13.0-beta1).

Any breaking changes are denoted with ⚠️.

### Removed

- ⚠️ Removed any `JoltPhysicsServer3D.G6DOF_*` enums that were previously (i.e. before Godot 4.3) not exposed by Godot itself. These can now be found as `PhysicsServer3D.G6DOF_*` instead.

### Changed

- ⚠️ Changed the way that the `body_test_motion` method of `PhysicsServer3D` discards contacts, which should fix issues related to jitter/ping-ponging. This affects `move_and_slide`, `move_and_collide` and `test_move`.
- ⚠️ Changed the inertia of shapeless bodies to be `(1, 1, 1)`, to match Godot Physics.
- Changed `SeparationRayShape3D` to not treat other convex shapes as solid, meaning it will now only ever collide with the hull of other convex shapes, which better matches Godot Physics.
- Mirrored the way in which angular limits are visualized for `JoltHingeJoint3D`, `JoltConeTwistJoint3D` and `JoltGeneric6DOFJoint`.

### Added

- Added support for `SoftBody3D`.
- Added support for double-precision.
- Added `compatibility_maximum` to `godot-jolt.gdextension` for controlling the maximum allowed Godot version. Note that changing this may expose you to subtle (or not so subtle) incompatibilities in the extension API. Change at your own risk.
- ⚠️ Added new project setting, "Use Enhanced Internal Edge Detection", which can help alleviate collisions with internal edges of `ConcavePolygonShape3D` and `HeightMapShape3D` shapes, also known as ghost collisions. This setting is enabled by default and may change the behavior of character controllers relying things like `move_and_slide` or `move_and_collide`.
- Added support for partial custom inertia, where leaving one or two components at zero will use the automatically calculated values for those specific components.
- Added error-handling for invalid scaling of bodies/shapes.
- Added project settings "Body Pair Cache Enabled", "Body Pair Cache Distance Threshold" and "Body Pair Cache Angle Threshold" to allow fine-tuning the scale by which collision results are reused inbetween physics ticks.
- Added the appropriate floating-point precision feature tag to `godot-jolt.gdextension` to prevent accidentally loading a single-precision build of Godot Jolt in a double-precision build of Godot and vice versa.

### Fixed

- ⚠️ Fixed issue with shape queries not returning the full contact manifold. This applies to the `collide_shape` method of `PhysicsDirectSpaceState3D` as well as the `body_test_motion` method of `PhysicsServer3D`, which subsequently affects the `test_move` and `move_and_collide` methods of `PhysicsBody3D` as well as the `move_and_slide` method of `CharacterBody3D`.
- ⚠️ Fixed issue with the `body_get_direct_state` method of `PhysicsServer3D` returning a non-null reference when the body has no space.
- Fixed issue with not being able to pass a physics space `RID` to `area_get_param`, `area_attach_object_instance_id` and `area_get_object_instance_id`.
- Fixed issue where the `inverse_inertia` property of `PhysicsDirectBodyState3D` would have some of its components swapped.
- Fixed issue where shapeless bodies wouldn't have custom center-of-mass applied to them.
- Fixed issue with high (>1) damping values producing different results across different update frequencies.
- Fixed issue where physics queries performed within the editor (e.g. editor plugins or tool scripts) would end up ignoring potentially large swathes of bodies in scenes with many physics bodies.
- Fixed crash when setting a negative `max_contacts_reported` on `RigidBody3D`.
- Fixed issue with certain `generic_6dof_joint_*` methods on `JoltPhysicsServer3D` not having the correct parameter names.
- Fixed issue where sleep timer would not reset when applying force/torque to a `RigidBody3D`.

<hr /><em>This discussion was created from the release <a href='https://github.com/godot-jolt/godot-jolt/releases/tag/v0.13.0-beta1'>0.13.0-beta1</a>.</em>

**Comment by Mantissa-23** [comment]
I was literally just rubber-ducking an issue I had with cylinder shapes not consistently returning the full contact manifold when my first-person character controller would perform a stairstep. This is amazing, thank you!

Is there a place I can drop a quick donation? You just saved me a lot of hair-pulling.

**Reply by mihe** [reply]
I'm glad to hear it came of some use so quickly! Frankly I should have caught that discrepancy when I first implemented the contact reporting.

> Is there a place I can drop a quick donation?

That won't be necessary, but I appreciate the offer. Consider donating to the [Godot Development Fund](https://fund.godotengine.org/) instead, if you aren't already.

**Comment by addmix** [comment]
Amazing. Beautiful.

With double-precision builds available, it reminds me of a discussion I had with jrouwe in the comments on a video of his, concerning double-precision velocities for space simulation games, avoiding instances of "the kraken" as seen in KSP, caused by the limited precision of velocities. Still not entirely sure if it would even be necessary with Jolt, but it could set apart Jolt as being the "do it all with no limits" type of physics engine.

What's the relationship of the max velocity options in projectsettings anyways? Is it only there as a failsafe, or does it have a tangible impact on simulation quality?

**Reply by mihe** [reply]
> What's the relationship of the max velocity options in projectsettings anyways? Is it only there as a failsafe, or does it have a tangible impact on simulation quality?

Mostly as a fail-safe, as I understand it. See jrouwe's answer [here](https://github.com/godot-jolt/godot-jolt/discussions/484#discussioncomment-6390918).

**Comment by ZenPyro** [comment]
**NOTE: I just updated my previous project from 4.2.1.stable.mono, to 4.3.beta.1.mono**

I am having a weird bug where sometimes when the player character's `CharacterBody3D` instance is falling, it will tell me I am passing an invalid transform to the `body_test_motion()` function. It does not seem to actually be affecting anything, but the error will start spewing out, and I cannot recreate it consistantly.

**ENTIRE ERROR:**
`E 0:00:34:0126   NativeCalls.cs:358 @ Godot.NativeInterop.godot_bool Godot.NativeCalls.godot_icall_0_40(nint, nint): body_test_motion failed due to being passed an invalid transform. A scale of (0.999960, 1.000000, 0.999960) is not supported by Godot Jolt for this shape type. Its scale will instead be treated as (1, 1, 1).
  <C++ Source>   src\spaces\jolt_physics_direct_space_state_3d.cpp:714 @ test_body_motion()
  <Stack Trace>  NativeCalls.cs:358 @ Godot.NativeInterop.godot_bool Godot.NativeCalls.godot_icall_0_40(nint, nint)
                 CharacterBody3D.cs:302 @ bool Godot.CharacterBody3D.MoveAndSlide()
                 Player.cs:371 @ void Player._PhysicsProcess(double)
                 Node.cs:2389 @ bool Godot.Node.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name&, Godot.NativeInterop.NativeVariantPtrArgs, Godot.NativeInterop.godot_variant&)
                 Node3D.cs:1071 @ bool Godot.Node3D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name&, Godot.NativeInterop.NativeVariantPtrArgs, Godot.NativeInterop.godot_variant&)
                 CollisionObject3D.cs:583 @ bool Godot.CollisionObject3D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name&, Godot.NativeInterop.NativeVariantPtrArgs, Godot.NativeInterop.godot_variant&)
                 PhysicsBody3D.cs:230 @ bool Godot.PhysicsBody3D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name&, Godot.NativeInterop.NativeVariantPtrArgs, Godot.NativeInterop.godot_variant&)
                 CharacterBody3D.cs:788 @ bool Godot.CharacterBody3D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name&, Godot.NativeInterop.NativeVariantPtrArgs, Godot.NativeInterop.godot_variant&)
                 Player_ScriptMethods.generated.cs:118 @ bool Player.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name&, Godot.NativeInterop.NativeVariantPtrArgs, Godot.NativeInterop.godot_variant&)
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(nint, Godot.NativeInterop.godot_string_name*, Godot.NativeInterop.godot_variant**, int, Godot.NativeInterop.godot_variant_call_error*, Godot.NativeInterop.godot_variant*)
`

**Reply by mihe** [reply]
I should perhaps rephrase that error message a bit. It used to actually fail (i.e. return) from `body_test_motion` when I first implemented the error, but I relaxed it a bit by changing it to just default the scaling to `(1, 1, 1)` instead, as mentioned in the message.

It should probably also mention that the error likely relates to `move_and_slide` or `move_and_collide`, which both rely on this `PhysicsServer3D.body_test_motion` function.

The error that's being reported is still somewhat real though, albeit a bit aggressive in its tolerance perhaps. You're likely building up enough numerical inaccuracies through whatever transformations you're doing to your `CharacterBody3D` (or its ancestor nodes) that its global transform is no longer represent a valid scaling for that particular shape type. I'm assuming your player uses a capsule shape, given the error message, in which case the scaling must be uniform on all axes.

You should be able to address it by doing this just before calling `MoveAndSlide()`:

```cs
Transform = Transform.Orthonormalized()
```

or if the error is coming from your ancestor nodes, this:

```cs
GlobalTransform = GlobalTransform.Orthonormalized()
```

I'll see if maybe the error tolerance can be relaxed a bit. I'd still like to keep the error check in there.

**Reply by ZenPyro** [reply]
AHHH! great! Thank you so much! 
It is most likely from my rotation code, here:
```c#
Vector3 curForwardDirTop = -GlobalTransform.Basis.Z.Normalized();
//NOTE: Redundant normalization here, but done for safety incase I change the ->
//-> earlier normalization code at somepoint and forget that the `directionVect` ->
//-> vector must be normalized here too.
Vector3 targetDirTop = directionVect.Normalized();
Vector3 crossDirTop = curForwardDirTop.Cross(targetDirTop).Normalized();

//Calculate the turning direction utilizing the already calculated cross-product.
//NOTE: The sign (positive or negative) of the cross-product determines if the ->
//-> perpendicular vector is up or down (since its on the `y-axis`), this also ->
//-> tells us if the shepherd is turning left or right.
turningDir = crossDirTop.Y;

float dotScalarTop = curForwardDirTop.Dot(targetDirTop);
float byAngleTop = Mathf.Acos(dotScalarTop);

if(byAngleTop > 0.001f && crossDirTop.Length() > 0.001){
Quaternion targetRotTop = new Quaternion(crossDirTop, byAngleTop * (float) delta * turnSpeed).Normalized();

Transform3D targetTransform = GlobalTransform;
targetTransform.Basis = new Basis(targetRotTop).Orthonormalized() * targetTransform.Basis.Orthonormalized();
GlobalTransform = targetTransform;
}
``` 

I do have one question though, I have always wondered this and this seems as good as time as any to ask it! The documentation for the `Orthonormalized()` instance function is: 

> "Returns the orthonormalized version of the basis matrix (useful to call occasionally to avoid rounding errors for orthogonal matrices)."

And I always took this as "don't use this every frame", as it could have some unintended affect or something. Is it safe to use `Orthonormalized()` every frame (since rotating happens most frames)? I only ask because why else would they make sure to mention "occasionally".

**Reply by mihe** [reply]
I'm not sure why it says that, to be honest. It is a moderately expensive function, involving 3 vector normalizations, so it's not something you want to be doing too often I guess. It could also be that constantly readjusting the basis vectors introduces some very slight rotational jitter/instability perhaps, but I'm anything but a math wiz, so don't quote me on that.

**Reply by ZenPyro** [reply]
Hahah! Your guess is as good as mine (probably better)! That being said, It might be from this function which I was not Orthonormalizing the returned (rotated) `Basis` instance:
```c#
GlobalBasis = GlobalBasis.Rotated(GlobalBasis.Y, yRot)
``` 

**UPDATE:** The error goes away completely with just the line `GlobalBasis = GlobalBasis.Rotated(GlobalBasis.Y, yRot).Orthonormalized();`. So, in my completely untested opinion, that's definitely what was causing it.

**Reply by mihe** [reply]
Out of curiosity, what type of shape(s) were you using for your `CharacterBody3D` when this error showed up? Did it have multiple shapes in it?

**Reply by ZenPyro** [reply]
So there was one capsule shape used as the singular `CollisionShape3D` instance for the `CharacterBody3D` instance. Then on top of that, there were a couple `Area3D` instances at differing sizes, but since they are `Area3D` Nodes, I doubt it had anything to do with it! I did have a neck turning behavior, but that only turned the mesh of the head, no actual physics bodies were used there.

***IMAGE OF SINGULAR `CharacterBody3D` COLLISION SHAPE:***
![image](https://github.com/godot-jolt/godot-jolt/assets/12946986/3b6bcb42-7eec-4409-a309-7b35f15aad89)


***IMAGE OF `Area3D` NODES:***
![image](https://github.com/godot-jolt/godot-jolt/assets/12946986/4dd265a7-2baa-4bee-addb-454af53cb0d0)

**Reply by mihe** [reply]
Hm, is the capsule shape offset in any way, in terms of its local transform? It's hard to tell from the image, but it looks like it's perhaps pushed up a bit on Y-axis?

**Reply by ZenPyro** [reply]
It is offset! Good eye, but I am curious as to why you think that would mess with physics?

***CAPSULE TRANSFORM:***
![image](https://github.com/godot-jolt/godot-jolt/assets/12946986/ad22e7df-1516-4bb9-8cc0-56bdbd1d48f4)

**Reply by mihe** [reply]
Any change to the shape's local transform (when you only have a single shape) causes it to internally be wrapped in what Jolt calls a `RotatedTranslatedShape`, which uses a different type of validity check for the scaling.

The reason I'm asking is because the scaling that was reported as being invalid for you should actually have passed just fine, so I'm trying to figure out what the actual underlying Jolt shape arrangement ended up being for you, but it sounds like it's just a regular `CapsuleShape` wrapped in a `RotatedTranslatedShape` with a non-zero translation, which unfortunately doesn't really explain this mysterious failure.

**Reply by ZenPyro** [reply]
Ahhh I see I see! If it might help, here's the actual shape's instance variables:
![image](https://github.com/godot-jolt/godot-jolt/assets/12946986/c22db3fc-f700-455b-a3d9-a379955d4836)

By the way I feel like I should just mention how insanely happy I am with how much you've done pushing the Godot engine forward with the Jolt Engine. I've always been on the Godot Engine, but your work has seriously improved so much! Please keep on going! I feel like you get asked this all the time, but is there any news on Jolt becoming implemented officially with Godot? Perhaps replacing the `GodotPhysics`? Have you been in touch with the physics team? I feel like you'd be such a valuable asset to them with the drive and dedication you have!

**Reply by mihe** [reply]
Could I maybe bother you to do one more thing? If you could open up the scene that contains your `CharacterBody3D` and click Project > Tools > Jolt Physics > Dump Debug Snapshots. That'll give you a prompt of where to save these snapshots, the path of which you'll find in the Output log when it's done generating it/them.

If you could zip that snapshot and upload it here (by dropping it in the comment box) I would appreciate it.

(The snapshots only contain the physics data for the scene you currently have focused in the editor, nothing else.)

**Reply by ZenPyro** [reply]
Okay sending it over now! Sorry I was working, wish I saw this sooner.

Here it is!
[jolt_snapshot_2024-06-09_17-24-42_175113.zip](https://github.com/user-attachments/files/15754236/jolt_snapshot_2024-06-09_17-24-42_175113.zip)

**Reply by ZenPyro** [reply]
I have one question that is kind of unrelated, but I went on a deep dive, and I feel like you're familiar with the engine. Since its completely off topic feel free not to answer haha! Its not so much a question, but a predicament? Two of the same lines of code, result in differing outputs.

Okay I'll show you a picture, and then I will break down the weird stuff:
![image](https://github.com/godot-jolt/godot-jolt/assets/12946986/73818d2f-9fe8-45e0-bbd7-a30bb535642f)

Now from looking at this, I want to ask you, what do you think the output would be? The same right? They are effectively the exact same print outs, but one was split up into a reference before calling the function `.Length()`. But extremely oddly, they do not result in the same value. Somehow they have opposite signs? Here is a picture of the print out:
![image](https://github.com/godot-jolt/godot-jolt/assets/12946986/23af9462-9680-4647-ae2d-f35d3519ac70)


This is so bafflingly odd, and I cannot seem to find a reason to why this happens, is it because Vector3 is a `struct`? This is concerning because when it comes to rotations, something could go terribly wrong if the signs just start being opposite!

**Reply by mihe** [reply]
It doesn't seem particularly odd to me. You're not negating the vector in the first one, you're negating the entire expression, including the `Length()` call.

**Reply by ZenPyro** [reply]
Ahh ahh! I missed that, thank you!

**Comment by DustStock** [comment]
I am having difficulty migrating a project from Godot 4.2 to Godot 4.3 and updating Jolt to this version.

Is there a process that I can follow somewhere to try to eliminate possible mistakes I am making?

If not, what is the easiest way to perform this migration?

**Reply by DustStock** [reply]
Well, I had downloaded the "double" release, but when I switched to the other one it worked fine. This was the error I was receiving:

>   core/extension/gdextension.cpp:1011 - No GDExtension library found for current OS and architecture (windows.x86_64) in configuration file: res://addons/godot-jolt/godot-jolt.gdextension

>   Failed loading resource: res://addons/godot-jolt/godot-jolt.gdextension. Make sure resources have been imported by opening the project in the editor at least once.

**Reply by mihe** [reply]
The "double" releases are meant to be used with double-precision builds of Godot (which you'd have to compile yourself) and are not compatible with the regular Godot releases, which are single-precision.

But yes, the error message you get when you have a mismatch of these floating-point precisions is not great. It's because I'm using Godot's "feature tags" in the `godot-jolt.gdextension` file to limit the binaries to their respective Godot builds, which means that when you have a mismatch then that version of Godot simply won't find any extension binaries to load, and gives you that cryptic error as a result.

If I didn't take this approach you'd just end up crashing instead.

**Comment by Cladis-dev** [comment]
Sorry for the noob question, but what does "symbol" mean on several of the .zip's?

**Reply by mihe** [reply]
They're debug symbols, which basically map the compiled machine code to the source code, so that you can do things like step through the code in a native debugger, or generate readable backtraces when crashing, etc.

**Comment by MaceWang2000** [comment]
![屏幕截图 2024-08-04 185024](https://github.com/user-attachments/assets/56cfb9fc-670a-4e9b-ba01-e327363d8fb0)
Honestly, I don't know how to describe the problem as far as what the picture above says. I removed all the physics bodies that had scaling issues, and it still happens as described above, I removed the _snap_up_stairs_check function as described above, and it's still a problem, the game works fine, it's just that it will keep getting errors, and after many troubleshooting attempts I've made, the problem does appear to be the jolt issue, not the godot version, so I hope the author can fix it! I hope the author can fix this problem.

Translated with DeepL.com (free version)

**Reply by mihe** [reply]
This has already been addressed in #925, and will be in the next release, which should hopefully be out some time in the next week or two.
